reach4help/reach4help
# feat: added partial protection route guard, used in find, new request

## Description

This PR enables the functionalities required for #906 and #907 by introducing a new route guard named `PartiallyProtectedRoute`. It can be used by using the `partiallyProtected` flag in the module definition instead of the `protected` flag.

The PR makes modifications to the `MasterPage` to use the new route guard if the module definition for a module enables it.

The PR also uses this new guard for the request module and uses it to determine if to show the authentication modal or not.

The goal of this new guard is to perform the checks necessary to determine if the user is authenticated and if the user has completed the onboarding procedure of verifying the phone number and adding the primary address.

The guard only determines this and adds it as a flag in the `auth` redux state. The guard does not perform any redirections or restrict access.

## Motivation

#906 and #907 required us to show the authentication modal with the appropriate step for completing onboarding, only when the user performs an operation that requires the user to be authenticated.

To show the modal only when the user performs an action that requires authentication when the user wasn't onboarded, the controllers must be able to determine if the user is authenticated and if the sure has been onboarded to the platform.

The process of determining if the user is onboarded or not requires a few steps and requests to be performed in sequence.

Rather than performing these steps and checks in every controller that requires this functionality, it is better to create it as a separate guard as controllers that require this functionality wouldn't be making use of the `protectedRoute` guard anyway.

## Testing Guidelines

1. Try to create an offer for a request or create a new request without being signed in, it should open a modal with options to sign in and you shouldn't be able to create the offer or the request.

2. Sign In/Sign up but don't verify your phone number and then try to create a new request or try to create an offer for an existing request. It should open up a modal asking you to enter a phone number and verify it.

3. Sign in/Sign up and verify your phone number but don't enter a primary address or fill in your profile information. Try to create a new request or make an offer for an existing request. It should open up a modal asking you to enter your address and complete your profile.

4. Sign in/Sign up with an account and complete all the onboarding and registration process. Try to create a new request or an offer for a request, the pop should not appear and instead the action must be performed

## Release Checklist


## Security Checklist

This PR creates, modifies, or deletes:




## Additional Notes
